File: credit-engine-api/src/main/java/com/credit/engine/application/dto/pricing/PricingParameterRequest.java
package com.credit.engine.application.dto.pricing;

import com.credit.engine.domain.model.receivable.ReceivableType;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;

import java.math.BigDecimal;
import java.time.LocalDate;

public record PricingParameterRequest(
        @NotNull(message = "O tipo de recebível é obrigatório")
        ReceivableType receivableType,

        @NotNull(message = "A taxa base é obrigatória")
        @DecimalMin(value = "0.0", message = "A taxa base não pode ser negativa")
        @DecimalMax(value = "100.0", message = "taxa base não pode ser maior que 100%")
        BigDecimal baseRate,

        @NotNull(message = "A taxa de spread é obrigatória")
        @DecimalMin(value = "0.0", message = "A taxa de spread não pode ser negativa")
        @DecimalMax(value = "100.0", message = "A taxa de spread não pode ser maior que 100%")
        BigDecimal spreadRate,

        @NotNull(message = "A data de vigência é obrigatória")
        LocalDate effectiveDate
) {
}

File: credit-engine-ui/src/services/pricing/dto/PricingParameterRequest.ts
import ValidationUtils from '../../../utils/ValidationUtils.ts';
import type { ReceivableType } from '../../receivable/ReceivableType.ts';

export default interface PricingParameterRequest {
    receivableType: ReceivableType | '';
    baseRate: number | null;
    spreadRate: number | null;
    effectiveDate: Date | null;
}

export const validatePricingParameterRequest = (request: PricingParameterRequest): { message: string }[] => {
    const errors: { message: string }[] = [];

    if (ValidationUtils.isBlank(request.receivableType))
        errors.push({ message: 'O tipo de recebível é obrigatório.' });

    if (request.baseRate === null || request.baseRate < 0 || request.baseRate > 100)
        errors.push({ message: 'A taxa base deve estar entre 0 e 100.' });

    if (request.spreadRate === null || request.spreadRate < 0 || request.spreadRate > 100)
        errors.push({ message: 'O spread deve estar entre 0 e 100.' });

    if (!request.effectiveDate)
        errors.push({ message: 'A data de vigência é obrigatória.' });

    return errors;
};
